swarm/network/stream/peer.go

```diff
@@ -136,73 +138,105 @@ func (p *Peer) Registrations() error {
 	if p.streamer.syncMode != SyncingAutoSubscribe {
 		return nil
 	}
-	timer := time.NewTimer(p.streamer.syncUpdateDelay)
-	select {
-	case <-timer.C:
-	case <-p.quit:
-		timer.Stop()
-		return nil
-	}
 
-	err := p.doRegistrations()
-	if err != nil {
-		return err
-	}
+	var change chan int
+
+	kad := p.streamer.delivery.kad
+	po := chunk.Proximity(p.bzzPeer.Over(), kad.BaseAddr())
+	newdepth := kad.NeighbourhoodDepth()
+	timer := time.NewTimer(p.streamer.syncUpdateDelay)
+	nn := po >= newdepth
 
 	for {
+		depth := newdepth
 		select {
-		case <-p.quit:
-			return nil
-		case <-p.bzzPeer.ChangeC:
-			// ugly hack here as I was getting double subscription requests in snapshot_sync_test,
-			// which means that new subscription requests were being issued before
-			// a first round finished and the servers were being created
-			//TODO: needs investigation about why that is
-			timer = time.NewTimer(p.streamer.syncUpdateDelay)
-			select {
-			case <-timer.C:
-			case <-p.quit:
-				timer.Stop()
-				return nil
+		case <-timer.C:
+			change = p.bzzPeer.ChangeC
+		case newdepth := <-change:
+			if changed := nn != (po >= depth); changed {
+				nn = !nn
+				if nn {
+					//request peer to subscribe to PO bins depth, depth+1, po - 1, po+1, ... MaxProxDisplay
+					poList := []int{depth, depth + 1, po - 1}
```

Comment:
this is not correct, since po can == depth etc.
You need 

```golang
					for i := depth; i <= kad.MaxProxDisplay; i++ {
						if po != i {
                                                     p.doSubscribe(i) // with some error handling
                                                }
					}
```